### ISSUE TITLE
Exception in msfconsole when using `portfwd` when remote host closes TCP connection

### ISSUE THREAD

**Issue by asoto-r7** [rapid7/metasploit-framework#9888]
While testing PR #9356, I stumbled across this bug:

TL;DR:  There appears to be an issue with port forwarding and connect.  I suspect that we need an exception handler at `tcp_client_channel.rb:90` to handle the socket being closed by the remote side, for example at the end of a successful HTTP request.

### Summary

From a Windows x64 meterpreter, forwarding TCP connections results in a traceback when the remote side closes the connection.

### Traceback(s)

I've seen two different tracebacks so far while recreating and debugging the issue.  When connecting with an external application, like `netcat`, I get:

```
msf5 > #<Thread:0x00007fe8637c4810@/Users/asoto/git/r7/metasploit-framework/lib/msf/core/thread_manager.rb:93 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	6: from /Users/asoto/git/r7/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
	5: from /Users/asoto/git/r7/metasploit-framework/lib/rex/thread_factory.rb:22:in `block in spawn'
	4: from /Users/asoto/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rex-core-0.1.13/lib/rex/io/socket_abstraction.rb:121:in `block in monitor_rsock'
	3: from /Users/asoto/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rex-core-0.1.13/lib/rex/io/socket_abstraction.rb:121:in `loop'
	2: from /Users/asoto/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rex-core-0.1.13/lib/rex/io/socket_abstraction.rb:188:in `block (2 levels) in monitor_rsock'
	1: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_client_channel.rb:90:in `close_write'
/Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_client_channel.rb:106:in `shutdown': stdapi_net_socket_tcp_shutdown: Operation failed: The handle is invalid. (Rex::Post::Meterpreter::RequestError)
```

When I connect inside msfconsole with `connect, I get:

```
#<Thread:0x00007fd754325680@/Users/asoto/git/r7/metasploit-framework/lib/msf/core/thread_manager.rb:93 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	17: from /Users/asoto/git/r7/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
	16: from /Users/asoto/git/r7/metasploit-framework/lib/rex/thread_factory.rb:22:in `block in spawn'
	15: from /Users/asoto/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rex-core-0.1.13/lib/rex/io/socket_abstraction.rb:121:in `block in monitor_rsock'
	14: from /Users/asoto/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rex-core-0.1.13/lib/rex/io/socket_abstraction.rb:121:in `loop'
	13: from /Users/asoto/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rex-core-0.1.13/lib/rex/io/socket_abstraction.rb:188:in `block (2 levels) in monitor_rsock'
	12: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_client_channel.rb:90:in `close_write'
	11: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_client_channel.rb:106:in `shutdown'
	10: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:172:in `send_request'
	 9: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:200:in `send_packet_wait_response'
	 8: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:133:in `send_packet'
	 7: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:823:in `to_r'
	 6: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:823:in `call'
	 5: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:595:in `to_r'
	 4: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:453:in `each'
	 3: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:453:in `each'
	 2: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:596:in `block in to_r'
	 1: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:323:in `to_r'
/Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:323:in `pack': no implicit conversion of nil into Integer (TypeError)
```

### Steps to reproduce

1. Get a meterpreter session (on a Windows 10 x64 host, in my case).

```
msf5 > sessions -l

Active sessions
===============

  Id  Name  Type                     Information                             Connection
  --  ----  ----                     -----------                             ----------
  1         meterpreter x64/windows  DESKTOP-Q05UKIU\root @ DESKTOP-Q05UKIU  192.168.108.1:54167 -> 192.168.108.217:4444 (192.168.108.217)
```

2. Use the built-in `connect` command to connect directly to `www.rapid7.com`.  Provide a basic HTTP header, which will return an `400 Bad Request` from the server.  This is fine.

```
[*] Connected to 52.85.208.191:80
GET / HTTP/1.1

HTTP/1.1 400 Bad Request
[ ... snip ... ]
```

3. Setup port forwarding to the `www.rapid7.com` web server.

```
msf5 > sessions -C 'portfwd add -r 52.85.208.191 -p 80 -l 8000'
[*] Running 'portfwd add -r 52.85.208.191 -p 80 -l 8000' on meterpreter session 1 (192.168.108.217)
[*] Local TCP relay created: :8000 <-> 52.85.208.191:80
```

4. Use the built-in `connect` command to test out the tunnel.  Provide a basic HTTP header, which will return an `400 Bad Request` from the server.  This will break and generate the second traceback.

5. In a separate window, run `nc 127.0.0.1 8000` to generate the first traceback.

**Comment by asoto-r7** [comment]
[This Meterpreter bug](https://github.com/rapid7/metasploit-payloads/issues/280) _may_ be related, in that we might be causing Meterpreter to crash when msfconsole generates an exception.  But I have no evidence to support/contradict that theory.  I just wanted to make sure both issues were linked.

**Comment by asoto-r7** [comment]
The second traceback (above) also occurs when using `exploit/windows/smb/psexec` against a Windows XP SP2 host (not domain-joined) without specifying credentials.  Note that the exploit is being forwarded through a routed tunnel.  The traceback does not occur when we are not forwarding traffic via the `route` command:

```
msf5> route add 172.16.104.0 255.255.255.0 1
[*] Route added
```

msf5 exploit(windows/smb/psexec) > set RHOST 172.16.104.128
msf5 exploit(windows/smb/psexec) > set PAYLOAD windows/meterpreter/bind_tcp
msf5 exploit(windows/smb/psexec) > options

Module options (exploit/windows/smb/psexec):

   Name                  Current Setting  Required  Description
   ----                  ---------------  --------  -----------
   RHOST                 172.16.104.128   yes       The target address
   RPORT                 445              yes       The SMB service port (TCP)
   SERVICE_DESCRIPTION                    no        Service description to to be used on target for pretty listing
   SERVICE_DISPLAY_NAME                   no        The service display name
   SERVICE_NAME                           no        The service name
   SHARE                 ADMIN$           yes       The share to connect to, can be an admin share (ADMIN$,C$,...) or a normal read/write folder share
   SMBDomain             .                no        The Windows domain to use for authentication
   SMBPass                                no        The password for the specified username
   SMBUser                                no        The username to authenticate as


Payload options (windows/meterpreter/bind_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  thread           yes       Exit technique (Accepted: '', seh, thread, process, none)
   LPORT     4444             yes       The listen port
   RHOST     172.16.104.128   no        The target address


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf5 exploit(windows/smb/psexec) > run

[*] Started bind handler
[*] 172.16.104.128:445 - Connecting to the server...
[*] 172.16.104.128:445 - Authenticating to 172.16.104.128:445 as user ''...

[-] 172.16.104.128:445 - FAILED! The remote host has only provided us with Guest privileges. Please make sure that the correct username and password have been provided. Windows XP systems that are not part of a domain will only provide Guest privileges to network logins by default.

#<Thread:0x00007fb306f228b0@/Users/asoto/git/r7/metasploit-framework/lib/msf/core/thread_manager.rb:93 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	17: from /Users/asoto/git/r7/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
	16: from /Users/asoto/git/r7/metasploit-framework/lib/rex/thread_factory.rb:22:in `block in spawn'
	15: from /Users/asoto/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rex-core-0.1.13/lib/rex/io/socket_abstraction.rb:121:in `block in monitor_rsock'
	14: from /Users/asoto/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rex-core-0.1.13/lib/rex/io/socket_abstraction.rb:121:in `loop'
	13: from /Users/asoto/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/rex-core-0.1.13/lib/rex/io/socket_abstraction.rb:188:in `block (2 levels) in monitor_rsock'
	12: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_client_channel.rb:90:in `close_write'
	11: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_client_channel.rb:106:in `shutdown'
	10: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:172:in `send_request'
	 9: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:200:in `send_packet_wait_response'
	 8: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet_dispatcher.rb:133:in `send_packet'
	 7: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:823:in `to_r'
	 6: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:823:in `call'
	 5: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:595:in `to_r'
	 4: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:453:in `each'
	 3: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:453:in `each'
	 2: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:596:in `block in to_r'
	 1: from /Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:323:in `to_r'
/Users/asoto/git/r7/metasploit-framework/lib/rex/post/meterpreter/packet.rb:323:in `pack': no implicit conversion of nil into Integer (TypeError)
```

**Comment by asoto-r7** [comment]
Alright, this was introduced with the migration to Ruby 2.5.0 which included, among other things, an upgrade to rex-socket 0.1.13:

The specific commit that introduced this bug is https://github.com/rapid7/metasploit-framework/commit/9559c06c10dbd773835a71bb380a3b9ce526ad11

**Comment by OJ** [comment]
Nice find @asoto-r7 ! It's a pain in the butt finding the source of issues like this, so thanks for putting the hard yards in. Any idea on what in rex-socket might be biting us in the butt?

**Comment by asoto-r7** [comment]
TLDR: Ruby 2.5.0 is the problem.  Help?

Alright, with @dmohanty-r7's help, we figured out that the move to Ruby 2.5.0 was where the trouble started.  Specifically, the following diff killed us:

`.ruby-version`:
```
@@ -1 +1 @@
-2.4.3
+2.5.0
```

`metasploit-framework.gemspec`:
```
@@ -163,7 +163,7 @@ Gem::Specification.new do |spec|
   # Library for parsing and manipulating executable binaries
   spec.add_runtime_dependency 'rex-bin_tools'
   # Rex Socket Abstraction Layer
-  spec.add_runtime_dependency 'rex-socket'
+  spec.add_runtime_dependency 'rex-socket', "0.1.13"
```

`Gemspec`:
```
@@ -18,19 +18,6 @@ group :development do
   gem 'pry'
   # module documentation
   gem 'octokit'
-  # Metasploit::Aggregator external session proxy
-  gem 'metasploit-aggregator' if [
-    'x86-mingw32', 'x64-mingw32',
-    'x86_64-linux', 'x86-linux',
-    'darwin'].include?(RUBY_PLATFORM.gsub(/.*darwin.*/, 'darwin'))
-  gem 'google-protobuf', "3.5.1" if [
-    'x86-mingw32', 'x64-mingw32',
-    'x86_64-linux', 'x86-linux',
-    'darwin'].include?(RUBY_PLATFORM.gsub(/.*darwin.*/, 'darwin'))
-  gem 'grpc', "1.8.3" if [
-    'x86-mingw32', 'x64-mingw32',
-    'x86_64-linux', 'x86-linux',
-    'darwin'].include?(RUBY_PLATFORM.gsub(/.*darwin.*/, 'darwin'))
 end
```

**Comment by busterb** [comment]
I don't think that was an 'upgrade' to rex-socket 0.1.13, but rather a pin to keep it at a specific version during the development of a PR. What's in master is still the same commit since 2016.

Looking at the nature of the backtrace, it appears this is failing due to string manipulation, possibly a hack from the Ruby 2.0 days to avoid updating Metasploit to use UTF-8 strings everwhere (this is why many files still start with # -*- coding: binary -*- too). The failing code in question hasn't changed much in 4 years either.

```
c94c2c1fd15 (skape         2005-04-09 20:45:09 +0000 316)   def to_r
3a4b5a02fb7 (hdm           2011-05-28 05:55:51 +0000 317)     # Forcibly convert to ASCII-8BIT encoding
3a4b5a02fb7 (hdm           2011-05-28 05:55:51 +0000 318)     raw = value.to_s.unpack("C*").pack("C*")
c94c2c1fd15 (skape         2005-04-09 20:45:09 +0000 319) 
242dd3c7dd9 (skape         2005-04-09 20:57:59 +0000 320)     if (self.type & TLV_META_TYPE_STRING == TLV_META_TYPE_STRING)
e6672398d36 (spoonm        2005-07-10 06:32:13 +0000 321)       raw += "\x00"
242dd3c7dd9 (skape         2005-04-09 20:57:59 +0000 322)     elsif (self.type & TLV_META_TYPE_UINT == TLV_META_TYPE_UINT)
242dd3c7dd9 (skape         2005-04-09 20:57:59 +0000 323)       raw = [value].pack("N")
e85b8adcf8c (Stephen Fewer 2010-09-14 23:23:15 +0000 324)     elsif (self.type & TLV_META_TYPE_QWORD == TLV_META_TYPE_QWORD)
255e792ed3c (hdm           2014-06-30 02:03:53 -0500 325)       raw = [ self.htonq( value.to_i ) ].pack("Q<")
242dd3c7dd9 (skape         2005-04-09 20:57:59 +0000 326)     elsif (self.type & TLV_META_TYPE_BOOL == TLV_META_TYPE_BOOL)
f43c0763131 (skape         2005-04-12 06:39:33 +0000 327)       if (value == true)
f43c0763131 (skape         2005-04-12 06:39:33 +0000 328)         raw = [1].pack("c")
f43c0763131 (skape         2005-04-12 06:39:33 +0000 329)       else
f43c0763131 (skape         2005-04-12 06:39:33 +0000 330)         raw = [0].pack("c")
f43c0763131 (skape         2005-04-12 06:39:33 +0000 331)       end
c94c2c1fd15 (skape         2005-04-09 20:45:09 +0000 332)     end
```

I'd suggest doing an inspection of the generated packet object in an exception handler around the generation of this packet. The contents might be surprising.

**Comment by busterb** [comment]
This is an interesting language change in 2.5 that might be related:

> Numerical comparison operators (<,<=,>=,>) no longer hide exceptions from #coerce method internally. Return nil in #coerce if the coercion is impossible. https://bugs.ruby-lang.org/issues/7688

**Comment by asoto-r7** [comment]
In `lib/rex/post/meterpreter/extensions/stdapi/net/socket_subsystem/tcp_client_channel.rb`:

>```
>100   def shutdown(how = 1)
>101     request = Packet.create_request('stdapi_net_socket_tcp_shutdown')
>102
>103     request.add_tlv(TLV_TYPE_SHUTDOWN_HOW, how)
>104     request.add_tlv(TLV_TYPE_CHANNEL_ID, self.cid)
>```

We're calling `shutdown()` and `self.cid` is `nil`.  So, we're trying to close a channel that doesn't exist, triggering an exception and traceback 1 above.

In what may be an unrelated issue, `self.cid` is a channel that doesn't exist, e.g. `5`, triggering an exception and traceback 2 above.

### Repeatability

Out of five runs, I encountered the following behaviors:
    1. worked fine
    2. Traceback 2: TypeError on `shutdown()` of `nil` channel
    3. Traceback 1: RequestError on `close_write` of invalid channel
    4. Traceback 1: RequestError on `close_write` of invalid channel
    5. Traceback 1: RequestError on `close_write` of invalid channel


### What definitely isn't a problem?
  - Ruby 2.5.0 and Ruby 2.4.3 both throw the same exception when attempting to convert `nil` into an integer via `[nil].pack("N")`.  So, it's not that this is a new exception with Ruby 2.5.0.
  - In fact, these exception is being pretty consistently generated by both the new ("bad") commit and the old ("good") commit.  But, somehow, the "good" commit is handling the exception somewhere.
